### ISSUE TITLE
SmallVector performance downgrade when small type on X86_64 

### ISSUE THREAD

**Issue by firedtoad** [llvm/llvm-project#58825]
Here is the test program
```
#include <benchmark/benchmark.h>
#include <llvm/ADT/SmallVector.h>
#include <vector>

template <typename V> static void BenchReserve(benchmark::State &state)
{
    auto sz = state.range(0);
    for (auto _ : state)
    {
        V v{};
        v.reserve(sz);
        for (auto i = 0; i < sz; i++)
        {
            v.emplace_back(i);
        }
        benchmark::DoNotOptimize(v);
    }
}

template <typename V> static void BenchReserveSmall(benchmark::State &state)
{
    auto sz = state.range(0);

    for (auto _ : state)
    {
        V v{};
        v.reserve(sz);
        for (auto i = 0; i < sz; i++)
        {
            v.emplace_back(i);
        }
        benchmark::DoNotOptimize(v);
    }
}
BENCHMARK_TEMPLATE(BenchReserve, std::vector<uint32_t>)->Range(65536, 65536);
BENCHMARK_TEMPLATE(BenchReserveSmall, llvm::SmallVector<uint32_t , 65536>)->Range(65536, 65536);
BENCHMARK_TEMPLATE(BenchReserve, std::vector<uint64_t>)->Range(65536, 65536);
BENCHMARK_TEMPLATE(BenchReserveSmall, llvm::SmallVector<uint64_t , 65536>)->Range(65536, 65536);

int main(int argc, char **argv)
{
    benchmark::Initialize(&argc, argv);
    benchmark::RunSpecifiedBenchmarks();
    return 0;
}
```
compile options
clang++-12  -fno-omit-frame-pointer -Wall -pedantic -Wextra -march=native  -fno-limit-debug-info -O2 -g -DNDEBUG -std=gnu++17 -MD -MT  -I/usr/lib/llvm-11/include 
Run output result
Run on (16 X 2593.91 MHz CPU s)
CPU Caches:
  L1 Data 48 KiB (x16)
  L1 Instruction 32 KiB (x16)
  L2 Unified 1280 KiB (x16)
  L3 Unified 43008 KiB (x16)
Load Average: 0.09, 0.03, 0.01
-------------------------------------------------------------------------------------------------------
Benchmark                                                             Time             CPU   Iterations
-------------------------------------------------------------------------------------------------------
BenchReserve<std::vector<uint32_t>>/65536                                     48142 ns         48135 ns        14538
BenchReserveSmall<llvm::SmallVector<uint32_t , 65536>>/65536     144439 ns       144411 ns       4846
BenchReserve<std::vector<uint64_t>>/65536                                      50970 ns        50961 ns        13730
BenchReserveSmall<llvm::SmallVector<uint64_t , 65536>>/65536      51382 ns        51373 ns        13628

we can see that ,when use uint32_t as type, performance downgrade so much
after reading the assembly code generated,i found the following line cause  the problem

```
template <class T>
using SmallVectorSizeType =
    typename std::conditional<sizeof(T) < 4 && sizeof(void *) >= 8, uint64_t,
                              uint32_t>::type;
```
when modify to this

```
using SmallVectorSizeType =
    typename std::conditional<sizeof(T) < 8 && sizeof(void *) >= 8, uint64_t,
                              uint32_t>::type;
```
problem gone.

**Comment by EugeneZelenko** [comment]
Could you please try 15 or `main` branch? 12 is too old.

**Comment by firedtoad** [comment]
with llvm-15 

`
clang++-15  -fno-omit-frame-pointer -Wall -pedantic -Wextra -march=native  -Wno-unused-parameter -Wno-unused-variable -Wno-ignored-qualifiers -fdiagnostics-color=auto  -fno-limit-debug-info -O2 -g -DNDEBUG CMakeFiles/adt_vec_ReserveSmall.dir/container/vector/Small.cpp.o -o adt_vec_ReserveSmall   -L/usr/local/lib  -Wl,-rpath,/usr/local/lib -fuse-ld=gold -lbenchmark -lpthread /usr/lib/llvm-15/lib/libLLVMSupport.a /usr/lib/llvm-15/lib/libLLVMCore.a /usr/lib/llvm-15/lib/libLLVMIRReader.a /usr/lib/llvm-15/lib/libLLVMAsmParser.a /usr/lib/llvm-15/lib/libLLVMBitReader.a /usr/lib/llvm-15/lib/libLLVMCore.a /usr/lib/llvm-15/lib/libLLVMBinaryFormat.a /usr/lib/llvm-15/lib/libLLVMRemarks.a /usr/lib/llvm-15/lib/libLLVMBitstreamReader.a /usr/lib/llvm-15/lib/libLLVMSupport.a -lrt -ldl -lpthread -lm /usr/lib/x86_64-linux-gnu/libz3.so /usr/lib/x86_64-linux-gnu/libz.so /usr/lib/x86_64-linux-gnu/libtinfo.so /usr/lib/llvm-15/lib/libLLVMDemangle.a
`
The Same Result
`
-------------------------------------------------------------------------------------------------------
Benchmark                                                             Time             CPU   Iterations
-------------------------------------------------------------------------------------------------------
BenchReserve<std::vector<uint32_t>>/65536                         48219 ns        48211 ns        14519
BenchReserveSmall<llvm::SmallVector<uint32_t , 65536>>/65536     149481 ns       149473 ns         4685
BenchReserve<std::vector<uint64_t>>/65536                         51082 ns        51070 ns        13694
BenchReserveSmall<llvm::SmallVector<uint64_t , 65536>>/65536      51232 ns        51224 ns        13681
`

**Comment by alexander-shaposhnikov** [comment]
Thanks for the report,
Out of curiosity I've run two experiments: 
1. use the modification suggested above (~branch perf/tune-small-vector)
2. use uint64_t unconditionally (on x86_64) (~branch perf/tune-small-vector-3) 
the results are available here: https://llvm-compile-time-tracker.com/index.php?config=NewPM-O3&stat=instructions%3Au&remote=alexander-shaposhnikov 
It feels like the current version is suboptimal / might be a wrong trade-off.
cc: @nikic

**Comment by nikic** [comment]
Is this related to the size of the type, or a strict aliasing problem (due to the size type and element being the same and thus allowed to alias)? For example, would using `float` exhibit the same performance characteristics or be faster?

**Comment by alexander-shaposhnikov** [comment]
I haven't checked what's going on in this specific case, but i suspect this might be related to the following: T* + uint32_t in general is not equivalent to pointer arithmetic (because of wrapping around UINT32_MAX).

**Comment by firedtoad** [comment]
Will it be fixed?

Alexander Shaposhnikov ***@***.***> 于2022年11月5日周六 16:32写道：

**Comment by alexander-shaposhnikov** [comment]
@firedtoad - does the issue reproduce with the latest Clang (built from sources) ?

**Comment by firedtoad** [comment]
I'll try it out

Alexander Shaposhnikov ***@***.***> 于2022年11月10日周四 07:55写道：

**Comment by firedtoad** [comment]
yes,it still threre when use the latest clang built from sources
